dig/graph.go

```diff
@@ -183,5 +183,38 @@ func (g *graph) registerConstructor(c interface{}) error {
 	}
 
 	g.nodes[objType] = &n
+
+	// object needs to be part of the graph to properly detect cycles
+	if cycleErr := g.detectCycles(&n); cycleErr != nil {
+		// if the cycle was detected delete from the graph
+		delete(g.nodes, objType)
```

Comment:
Got it, that does introduce a risk when this is executed in parallel. You might have to lock cycle detection function to only one execution at a time. One alternative to getting around that is treat the root node specially, on line 212 check g.nodes[dep] or root, check for cycles.